storm-server/src/main/java/org/apache/storm/scheduler/blacklist/BlacklistScheduler.java

```diff
@@ -129,7 +142,34 @@ public void schedule(Topologies topologies, Cluster cluster) {
         return underlyingScheduler.config();
     }
 
-    private void badSupervisors(Map<String, SupervisorDetails> supervisors) {
+    private Map<String, Integer> getNodeAssignmentFailures() {
```

Comment:
In the original backlist scheduler, when we decide to blacklist a node or not, we have `blacklist.scheduler.tolerance.time.secs` and `blacklist.scheduler.tolerance.count`. 

If the node has heartbeat failure more than the `count` in the `tolerance.time.secs` period, it will be blacklisted. 

But here with getNodeAssignmentFailures, since it's using a `RotatingMap` with 3 buckets and it rotates every 3600s (1hr), which means it keeps 2~3 hrs of data in the RotatingMap.  So `getNodeAssignmentFailures` here means if sending nodeAssignment to a node failed more than `blacklist.scheduler.assignment.failure.tolerance.count` times, it will be added to the blacklist. 

This is not consistent with heartbeatFailures and can be confusing. It is some kind of failure tolerance but doesn't follow the same logic and doesn't not use the same `blacklist.scheduler.tolerance.time.secs`. 

I think we should follow the same logic to determine whether blacklisting a node or not.

(The resume part is the same, which is good. It uses the same logic as the blacklist caused by heartbeat failure.)

Reply:
My intent on the rotating map was to drop supervisors over time that have potentially left the cluster.

In my mind, if sending assignments fails every time you talk to a supervisor regardless of how far back, it means you have a suspicious supervisor.  But I'm fine with tracking the exact times of the events to see if they fall within a tolerance window if you prefer.

Instead of tracking a count, I could track a list of failure times and drop the old failures if they're before the tolerance time limit?  Would that do what you desire?

Reply:
My idea is to take a snapshot of <Node, AssignementFailureCount> in `scheduler.schedule` and put it into another `badSupervisorsToleranceSlidingWindow`. And then clean up the map.  In this way, we can follow the same logic as bad supervisors caused by hearbeatFailure